# C14U5: Anthropology

In the following notebook, students will be able to:
* Create Python functions to process datasets using pandas
* Make a distinction between function definitions and function calls.
* Write code that accepts user input.



## 🟢 Module 1: Restaurant Culture Demographics Analysis


You will be given a dataset that provides restaurant inspections, violations, grades and adjudication information. Food is important in various cultures, and here we explore the types of food that each borough in NYC are exposed to.

First, use the pandas library to read in the data file.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Unit 5 - Anthropologie/restaurant_data_set_2.csv - restaurant_data_set_2.csv.csv")

In [ ]:
#Display the first 5 rows of the dataset
df.head()

CAMIS  ... GRADE
0  41338311  ...   NaN
1  50055587  ...     A
2  50046026  ...   NaN
3  41576311  ...     A
4  50000641  ...   NaN

[5 rows x 9 columns]

In [ ]:
#Find how many rows the dataset contains.
print(len(df))

404249


**Review**: Use the .loc[] attribute to filter the dataset to only show the information for each borough of NYC. Store each borough's information in its own variable.

In [ ]:
#Brooklyn
brooklyn = df.loc[df["BORO"] == "Brooklyn"]

In [ ]:
#Queens
queens = df.loc[df["BORO"] == "Queens"]

In [ ]:
#Manhattan
manhattan = df.loc[df["BORO"] == "Manhattan"]

In [ ]:
bronx = df.loc[df["BORO"] == "Bronx"]

In [ ]:
#Staten Island
staten = df.loc[df["BORO"] == "Staten Island"]

### Investigate
Fast food is a convenience such that you can grab it and go quickly on a work day and it is cheap. The downside is that it is unhealthy. Yet somehow many low-income and food-insecure households rely on fast-food restaurants. 

Write a program that finds the percentage of each of the following fast-food restaurants in each borough.
* Mcdonald's 
* Burger King
* Wendy's 
* Popeyes

In [ ]:
def Mcdonalds(borough):
  Mcdonalds = len(borough.loc[borough["DBA"] == "MCDONALD'S"])
  total = len(borough)
  print(Mcdonalds/total *100)
  return (Mcdonalds/total *100)

borough_list = [brooklyn, queens, manhattan, bronx, staten]

for i in borough_list:
  print(Mcdonalds(i)) 

0.42571473207311356
0.42571473207311356
0.5955662202622221
0.5955662202622221
0.3965652031372153
0.3965652031372153
1.2032660077352815
1.2032660077352815
0.6530550968323074
0.6530550968323074


In [ ]:
def burger_king(borough):
  burger_king = len(borough.loc[borough["DBA"] == "BURGER KING"])
  total = len(borough)
  print(burger_king/total *100)
  return (burger_king/total *100)

borough_list = [brooklyn, queens, manhattan, bronx, staten]

for i in borough_list:
  print(burger_king(i)) 

0.39544602405874085
0.39544602405874085
0.31453678783358735
0.31453678783358735
0.0788086651703528
0.0788086651703528
0.5720885259991405
0.5720885259991405
0.5029274883651103
0.5029274883651103


In [ ]:
def wendys(borough):
  wendys = len(borough.loc[borough["DBA"] == "WENDY'S"])
  total = len(borough)
  print(wendys/total *100)
  return (wendys/total *100)

borough_list = [brooklyn, queens, manhattan, bronx, staten]

for i in borough_list:
  print(wendys(i)) 

0.09666458365880332
0.09666458365880332
0.12430147972804999
0.12430147972804999
0.08763523566943232
0.08763523566943232
0.21755479157713797
0.21755479157713797
0.2552169343942351
0.2552169343942351


In [ ]:
def popeyes(borough):
  popeyes = len(borough.loc[borough["DBA"] == "POPEYES"])
  total = len(borough)
  print(popeyes/total *100)
  return (popeyes/total *100)

borough_list = [brooklyn, queens, manhattan, bronx, staten]

for i in borough_list:
  print(popeyes(i)) 

0.25581940321824714
0.25581940321824714
0.2950809040500665
0.2950809040500665
0.11852823241621062
0.11852823241621062
0.6768371293510959
0.6768371293510959
0.24020417354751541
0.24020417354751541


Change your code into a function that can produce percentages based on what restaurant you are searching the dataframe for. 

In [ ]:
def restaurant_type(restaurant):
  borough_list = [brooklyn, queens, manhattan, bronx, staten]
  for i in borough_list:
    restaurant = len(i.loc[i["DBA"] == restaurant])
    total = len(i)
    print(restaurant/total *100)
    return (restaurant/total *100)

In [ ]:
#Don't forget to call your function!
restaurant_list = ["MCDONALD'S", "BURGER KING", "WENDY'S", "POPEYES"]

counter = 0

for i in restaurant_list:
  print(restaurant_type(i), restaurant_list[counter])
  counter += 1

0.42571473207311356
0.42571473207311356 MCDONALD'S
0.39544602405874085
0.39544602405874085 BURGER KING
0.09666458365880332
0.09666458365880332 WENDY'S
0.25581940321824714
0.25581940321824714 POPEYES


Each borough has the following demographics (taken from US census data): <br><br>

Brooklyn
* White - 49.8%
* Black - 33.8%
* Native American - 0.9% 
* Asian - 12.7%
* Native Hawaiian and other Pacific Islander - 0.1%
* Mixed - 2.7%
* Hispanic or Latino - 18.9% 
<br><br>

Queens
* White - 47.8%
* Black - 20.7%
* Native American - 1.3% 
* Asian - 26.9%
* Native Hawaiian and other Pacific Islander - 0.2%
* Mixed - 3.0%
* Hispanic or Latino - 28.2%
<br><br>

Manhattan
* White - 64.6%
* Black - 17.8%
* Native American - 1.2%
* Asian - 12.8%
* Native Hawaiian and other Pacific Islander - 0.2% 
* Mixed - 3.4%
* Hispanic or Latino - 25.6% 
<br><br>

Bronx
* White -  44.7%
* Black - 24.3%
* Native American - 0.4% 
* Asian - 14.1%
* Native Hawaiian and other Pacific Islander - 0.1% 
* Mixed - 3.6%
* Hispanic or Latino - 29.1% 
<br><br>

Staten Island
* White - 74.5%
* Black - 11.6%
* Native American - 0.7% 
* Asian - 10.9%
* Native Hawaiian and other Pacific Islander - 0.1% 
* Mixed - 2.2%
* Hispanic or Latino - 18.6% 


Questions:
1. How can we narrow down the data further to find how different racial/ethnic groups are impacted by its population of fast-food restaurants? (by zip code)

2. How may the connection between fast-food restaurants and black and low-income neighborhoods contribute to the causes of the obesity epidemic in these poulations?

3. What can we do with computer science to help change this?

# 🟢 Module 2: Language Translation Function

In 1971, a linguist named Morris Swadesh published a list of words most resistant to cross-cultural change. The list, now referred to as the Swadesh list (there are two primary versions, of 100 and 200 words), has been used to reconstruct genealogical relationships among the languages in a language family.

AMNH colleagues, Dr. Whiteley and Dr. Ward Wheeler (Curator, Invertebrate Zoology) have been studying historical relationships among languages, beginning with Uto-Aztecan languages of Middle and North America, using genetic methods developed in the field of evolutionary biology. With phylogenetic software, they systematically compare cognate words as sequences of sounds (phonemes). Where available, they also analyze grammatical (morphological and syntactic) data.

**TASK**: Use the data given to write a program that accepts user-input and translates a given English word using the Uto-Aztecan language. Note: Not every English word is recorded, so be sure to create a conditional for the case that it does not exist in the dataset.

In [ ]:
def addition(addend1, addend2):
  y = addend1 + addend2
  return y

input1 = int(input("enter first number"))
input2 = int(input("enter second number"))

addition(input1, input2)

enter first number3
enter second number5


8

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Unit 5 - Anthropologie/uto_aztecan_languages_datasets (1).csv")

In [ ]:
df.head()

Language  ...                 Unnamed: 204
0                               NaN  ...                         tipa
1                               NaN  ...                          NaN
2                   Northern_Paiute  ...               t\=Wt\super si
3                      Western_Mono  ...   inott\super sPtt\super siP
4  Tümpisa_Shoshone_(Panamint,Koso)  ...  tWtWtt\super sitt\super si 

[5 rows x 205 columns]

In [ ]:
df.columns

Index(['Language', 'louse', 'two', 'water', 'ear', 'die', 'I', 'liver', 'eye',
       'hand',
       ...
       'many', 'all', 'know', 'cloud', 'swim', 'belly ', 'big', 'hot',
       'lie_(down)', 'small'],
      dtype='object', length=103)

In [ ]:
for i in df.columns:
  if "Umnamed" in i:
    df = df.drop([i], axis=1)

In [ ]:
df.head(5)

Language      louse  ...     lie_(down)          small
0                               NaN        IPA  ...            IPA            IPA
1                               NaN        NaN  ...            NaN            NaN
2                   Northern_Paiute  potsiɑppi  ...  hɑpi | ­pokwɑ         tɯ̄tsi
3                      Western_Mono    pusi̱ʔɑ  ...          hɑpi­  inottsiɁttsiɁ
4  Tümpisa_Shoshone_(Panamint,Koso)  posiɑttsi  ...           hɑpi  tɯtɯttˢittˢi 

[5 rows x 103 columns]

In [ ]:
possible_words = list(df.columns)
possible_words.remove("Language")
print(possible_words)

['louse', 'two', 'water', 'ear', 'die', 'I', 'liver', 'eye', 'hand', 'hear', 'tree', 'fish', 'name_(noun,though_some_overlap_with_verb)', 'stone', 'tooth', 'breast', 'you_(sing.)', 'path', 'bone', 'tongue', 'skin', 'night', 'leaf', 'rain_(vb.,mostly)', 'kill_(pl._obj.)', 'kill_(sing./dual_obj.)', 'blood', 'horn', 'person', 'knee', 'one', 'nose', 'full', 'come', 'star', 'mountain', 'fire', 'we', 'to_drink', 'to_see', 'bark(tree)', 'new', 'dog', 'sun', 'to_fly', 'fat,_grease_(substance)', 'moon', 'give', 'heart', 'feather_(not_down)', 'white', 'yellow', 'bird', 'head', 'earth_1_(primarily_mud_clay)', 'earth_2_(primarily_land _landform_ground_soil_world)', 'foot', 'black', 'mouth', 'green_blue', 'sleep', 'what', 'root', 'claw', 'bite', 'ash/ashes', 'red', 'eat', 'egg', 'who', 'dry', 'hair', 'smoke_(primarily_noun)', 'not', 'this', 'seed', 'woman', 'round', 'long', 'stand', 'good', 'man', 'cold', 'flesh_(meat)', 'neck', 'to_say', 'to_burn', 'tail', 'sand', 'that/that_one', 'walk', 'sit', '

In [ ]:
possible_languages = df["Language"].tolist()
print(possible_languages)

[nan, nan, 'Northern_Paiute', 'Western_Mono', 'Tümpisa_Shoshone_(Panamint,Koso)', 'Big_Smokey_Valley_Shoshone', 'Western_Shoshone', 'Shoshone', 'Comanche', 'Kawaiisu', 'Chemehuevi', 'Southern_Paiute', 'Southern_Ute', 'Orayvi_Hopi', 'Pahkannil_(Tübatulabal)', 'Kitanemuk', 'Serrano', 'Tongva_(Gabrielino)', 'Luiseño', 'Acjachemen_(Juaneño)', 'Cupeño', 'Cahuilla', "Tohono_O'odham", 'Pima_Bajo', 'N_Tepehuan', 'S_Tepehuan', 'Guarijío', 'Rarámuri', 'Eudeve', 'Ópata', 'Arizona_Yoeme_(Yaqui)', 'Mayo', 'Tubar', 'Cora', 'Wixarika_(Huichol)', 'Classical_Nahuatl', 'Tetelcingo_Mexicano', 'Pochutla_Mexicano', 'Pipil', nan, 'Ipai', 'Tewa', 'Zuni']


In [ ]:
def translate(dataframe, language, enlglishWORD):
  translation = df.loc[df["Language"] == language, englishWORD]
  translation = translation.to_string(index=False)
  print(translation)
  return translation

In [ ]:
language_choice = input("Enter name: ")
word_choice = input("Enter word: ")

translate(df, language_choice, word_choice)

Enter name: Western Mono
Enter word: pusi̱ʔɑ	


NameError: ignored

# 🟢 Module 3: Error Rates in Users of Automatic Face Recognition Software

While facial recognition sounds nice and may have its perks, there is bias in all algorithms as long as there is a lack of diversity in its creators. If not made correctly, facial recognition risks being weaponized by law enforcement against marginalized communities around the world. From New Delhi to New York, this invasive technology can turn our identities against us and can undermine human rights. 

Use the given dataset to create a plot of the error rates based on gender and race. You will be given an excel file and are responsible for saving it as a csv file to use.